File: src/test/java/com/running/dto/buddy/test/java/cases/validate/BaseClassValidationTest.java
package com.running.dto.buddy.test.java.cases.validate;

import com.runninglane.dto.buddy.DtoBuddy;
import com.runninglane.dto.buddy.exception.DtoBuddyBadInputException;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

interface PrivateInterface {
}

public class BaseClassValidationTest {
    protected interface ProtectedInterface {
    }

    public interface PublicInterface {
    }

    class FinalClassWithNoProperties {
    }

    class FinalClassWithImmutableProperties {
        private final String name = "John";

        public String getName() {
            return name;
        }
    }

    class FinalClassWithMutableProperties {
        private String name = "John";

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
    }

    public static class OpenClass {
    }

    @Test
    public void shouldFailWhenClassIsProtected() {
        assertThrows(DtoBuddyBadInputException.class, () ->
            DtoBuddy.implementor().withBaseClass(ProtectedInterface.class).implement());
    }

    @Test
    public void shouldFailWhenClassIsPrivate() {
        assertThrows(DtoBuddyBadInputException.class, () ->
            DtoBuddy.implementor().withBaseClass(PrivateInterface.class).implement());
    }

    @Test
    public void shouldFailWhenClassIsFinalAndIsNotCompleteAndMutable() {
        assertThrows(DtoBuddyBadInputException.class, () ->
            DtoBuddy.implementor(FinalClassWithImmutableProperties.class).implement());
    }

    @Test
    public void shouldFailWhenClassIsFinalAndIsCompleteAndMutable() {
        assertDoesNotThrow(() ->
            DtoBuddy.implementor(FinalClassWithNoProperties.class).implement());
        assertDoesNotThrow(() ->
            DtoBuddy.implementor(FinalClassWithMutableProperties.class).implement());
    }

    @Test
    public void shouldSuccessWhenClassIsOpenAndPublic() {
        assertDoesNotThrow(() ->
            DtoBuddy.implementor().withBaseClass(OpenClass.class).implement());

        assertDoesNotThrow(() ->
            DtoBuddy.implementor().withBaseClass(PublicInterface.class).implement());
    }

}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/validate/BaseClassValidationTest.kt
package com.runninglane.dto.buddy.test.cases.validate

import com.runninglane.dto.buddy.DtoBuddy
import com.runninglane.dto.buddy.exception.DtoBuddyBadInputException
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows
import kotlin.test.Test

private interface PrivateInterface

open class BaseClassValidationTest {

    protected interface ProtectedInterface
    internal interface InternalInterface
    interface PublicInterface
    class FinalClass
    open class OpenClass

    @Test
    fun shouldFailWhenClassIsProtected() {
        assertThrows<DtoBuddyBadInputException> {
            DtoBuddy.implement(ProtectedInterface::class.java)
        }
    }

    @Test
    fun shouldFailWhenClassIsPrivate() {
        assertThrows<DtoBuddyBadInputException> {
            DtoBuddy.implement(PrivateInterface::class.java)
        }
    }

    @Test
    fun shouldFailWhenClassIsFinal() {
        assertThrows<DtoBuddyBadInputException> {
            DtoBuddy.implement(FinalClass::class.java)
        }
    }

    @Test
    fun shouldSuccessWhenClassIsOpenAndPublic() {
        assertDoesNotThrow {
            DtoBuddy.implement(OpenClass::class.java)
        }

        assertDoesNotThrow {
            DtoBuddy.implement(PublicInterface::class.java)
        }

        assertDoesNotThrow {
            // Public internal
            DtoBuddy.implement(InternalInterface::class.java)
        }
    }

}